src/Microsoft.TemplateEngine.Cli/Commands/create/InstantiateCommand.TabCompletion.cs

```diff
@@ -110,5 +118,78 @@ protected internal override IEnumerable<CompletionItem> GetCompletions(Completio
                 yield return completion;
             }
         }
+
+        private static IEnumerable<CliTemplateInfo> GetAllowedTemplates(TemplateConstraintManager constraintManager, TemplateGroup templateGroup)
+        {
+            if (templateGroup.Templates.SelectMany(t => t.Constraints).Any())
+            {
+                CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+                cancellationTokenSource.CancelAfter(ConstraintEvaluationTimeout);
+                var constraintEvaluationTask = templateGroup.GetAllowedTemplatesAsync(constraintManager, cancellationTokenSource.Token);
+                Task.Run(async () =>
+                {
+                    try
+                    {
+                        await constraintEvaluationTask.WaitAsync(cancellationTokenSource.Token).ConfigureAwait(false);
+                    }
+                    catch (TaskCanceledException)
+                    {
+                        //do nothing
+                    }
+                }).GetAwaiter().GetResult();
+
+                if (constraintEvaluationTask.IsCompletedSuccessfully)
+                {
+                    return constraintEvaluationTask.Result;
+                }
+            }
+            return templateGroup.Templates;
```

Comment:
I'm probably bit lost - but it seems we return all templates here, not just the allowed ones returned by `GetAllowedTemplatesAsync` - why is that?

Reply:
we return all the templates in the group when:
- no constraints are defined
- constraints are defined, but evaluation didn't complete in time.

If the constraints are not evaluated within the timeout, the template in a group will be accepted for completion. Worst case scenario: the template won't run when instantiating.